socrata/resource.py

```diff
@@ -167,16 +169,28 @@ def show(self, uri):
             auth = self.auth
         ))
 
-
     def _wait_for_finish(self, is_finished, is_failed, progress, timeout, sleeptime):
-        started = time.time()
-        while not is_finished(self):
-            current = time.time()
-            if timeout and (current - started > timeout):
-                raise TimeoutException("Timed out after %s seconds waiting for completion for %s" % (timeout, str(self)))
-            me = self.show()
-            progress(self)
-            if is_failed(self):
-                raise ResourceFailedException(me)
-            time.sleep(sleeptime)
-        return self
+        def retry(e, attempts):
+            if attempts < 5:
+                attempts = attempts + 1
+                sleep(attempts * attempts * 2)
+                return wff(time.time(), attempts)
+            else:
+                raise e
+
+        def wff(started, attempts):
+            while not is_finished(self):
+                current = time.time()
+                if timeout and (current - started > timeout):
+                    timeoutWarn = "Timed out after %s seconds waiting for completion for %s on attempts %i" % (timeout, str(self), attempts)
+                    log.warn(timeoutWarn)
+                    ex = TimeoutException(timeoutWarn)
+                    retry(e, attempts)
+                me = self.show()
```

Comment:
This is the line we actually want to guard with a try/catch -- the actual HTTP request is made here and can fail with a timeout exception from the HTTP client, or other `requests.exceptions.RequestException` - we basically want to catch all of those, plus our own `UnexpectedResponseException` (because we want to retry on 5xx responses too) like we do over in [sources.py](https://github.com/socrata/socrata-py/blob/master/socrata/sources.py#L167-L175).

So something like:
```python
consecutive_failures = 0
last_exception = None
while not is_finished(self):
  # check consective_failures isn't too high, if it is throw last_exception
  # check the user-requested timeout and throw TimeoutException if it's taken too long
  try:
    me = self.show()
  except RequestException as e:
    last_exception = e
    consecutive_failures += 1
    continue
  except UnexpectedResponseException:
    if 500 <= e.status <= 599:
      last_exception = e
      consecutive_failures += 1
      continue
    raise e
  consecutive_failures = 0
  progress(self)
  if is_failed(self):
    throw ResourceFailed
  time.sleep(zzz)
return self
```